### PR TITLE
StaticSite module cleanup and optional CloudFront disablement for development.

### PR DESCRIPTION
**Pull request by voodooGQ**
This PR presents some general cleanup of the StaticSite module as well as making the CloudFront portion of the deployment optional for development purposes.

An integration test has been provided to verify the StaticSite module works as intended, however, due to the amount of time that would need to be taken to launch CloudFront that has been disabled for the test.

Additional documentation provided outlining the length of time that may come from running CloudFront with the build.

### CODE REVIEW

**Comment by voodooGQ** [comment]
Working on the Travis fixes. IDE is not cooperating. Will notify when ready.

**Comment by voodooGQ** [comment]
Ready for review

**Comment by voodooGQ** [comment]
After speaking with @troyready we are going to make the files launched with CloudFront disabled as publically accessible so they can be seen from the bucket.

File: docs/source/module_configuration/staticsite.rst [hunk 1]
```diff
@@ -38,6 +38,16 @@ be displayed on stack creation/updates.
 A start-to-finish example walkthrough is available
 in the :ref:`Conduit quickstart<qs-conduit>`.
 
+**Please note:** The CloudFront distribution will take a significant amount
```
Comment (voodooGQ) [inline docs/source/module_configuration/staticsite.rst]: Should we also include a prompt informing at launch time?

File: src/runway/blueprints/staticsite/staticsite.py [hunk 2]
```diff
@@ -100,57 +103,248 @@ class StaticSite(Blueprint):  # pylint: disable=too-few-public-methods
...
+
+    def add_bucket(self):
+        """Add the bucket resource along with an output of it's name.
+
+        Returns:
+            dict: The bucket resource
+
+        """
+        variables = self.get_variables()
+        access = s3.PublicRead if (variables['CFDisabled'] == 'true') else s3.Private
```
Comment (troyready) [inline src/runway/blueprints/staticsite/staticsite.py]: This should be a CloudFormation If (so it's evaluated at stack deployment time) on AccessControl, not a python if.

Will look something like `AccessControl=If('CFEnabled', s3.Private, s3.PublicRead)`

File: src/runway/blueprints/staticsite/staticsite.py [hunk 3]
```diff
@@ -198,7 +410,13 @@ def create_template(self):
...
+    def add_index_rewrite_role(self):
+        """Add an index rewrite role to the template.
+
+        Return:
+            dict: The index rewrite role
+        """
+        return self.template.add_resource(
             iam.Role(
                 'CFDirectoryIndexRewriteRole',
                 Condition='DirectoryIndexSpecified',
```
Comment (troyready) [inline src/runway/blueprints/staticsite/staticsite.py]: Might as well make this the 'CFEnabledAndDirectoryIndexSpecified' condition as well

File: src/runway/blueprints/staticsite/staticsite.py [hunk 4]
```diff
@@ -100,57 +103,248 @@ class StaticSite(Blueprint):  # pylint: disable=too-few-public-methods
...
+        index_rewrite_version = self.add_cloudfront_directory_index_rewrite_version(
+            index_rewrite
+        )
+        lambda_function_associations = self.get_lambda_associations(index_rewrite_version)
+        distribution_options = self.get_cloudfront_distribution_options(
+            bucket,
+            oai,
+            lambda_function_associations
+        )
+        distribution = self.add_cloudfront_distribution( # noqa pylint: disable=unused-variable
```
[the remark this answers is not in the document]
Comment (voodooGQ) [reply src/runway/blueprints/staticsite/staticsite.py]: Attempting to keep the consistency with the rest of the method. All other variables have an assignment. Willing to remove if you feel strongly.

File: src/runway/blueprints/staticsite/staticsite.py [hunk 5]
```diff
@@ -77,6 +77,9 @@ class StaticSite(Blueprint):  # pylint: disable=too-few-public-methods
                     'default': '',
                     'description': '(Optional) Domain aliases the '
                                    'distribution'},
+        'CFDisabled': {'type': CFNString,
```
[the remark this answers is not in the document]
Comment (voodooGQ) [reply src/runway/blueprints/staticsite/staticsite.py]: Back from the Future. 10-4

File: src/runway/blueprints/staticsite/staticsite.py [hunk 6]
```diff
@@ -100,57 +103,248 @@ class StaticSite(Blueprint):  # pylint: disable=too-few-public-methods
...
+                PolicyDocument=Policy(
+                    Version="2012-10-17",
+                    Statement=[
+                        Statement(
+                            Effect=Allow,
+                            Principal=Principal('*'),
+                            Action=[Action('s3', 'getObject')],
+                            Resource=[
+                                Join('', ['arn:aws:s3:::',
+                                          bucket.ref(),
```
[the remark this answers is not in the document]
Comment (voodooGQ) [reply src/runway/blueprints/staticsite/staticsite.py]: I've tried many variations of this (example: `GetAtt(bucket, 'Arn')` & `bucket.get_att('Arn') in the main array) but the system fails on it each time. Doesn't really give a great error message either, just that the Policy failed: 

image

Any ideas? I thought you needed the bucket name on a bucket policy (at least that's what I've seen from examples)

File: docs/source/module_configuration/staticsite.rst (hunk 1 above)
Comment (voodooGQ) [reply docs/source/module_configuration/staticsite.rst]: Moved this to the `deploy` method as it's not changing the YAML and should probably be out of the `pre_build` hooks.

File: src/runway/blueprints/staticsite/staticsite.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (troyready) [reply src/runway/blueprints/staticsite/staticsite.py]: I'm saying this:

```
                                Join('', ['arn:aws:s3:::',
                                          bucket.ref(),
                                          '/*'])
```

should become:
```
                                Join('', [bucket.get_att('Arn'),
                                          '/*'])
```

It's either that, or to keep your Ref, the technically correct thing to do would be to swap out the string "aws" for a reference to the [partition psuedoparameter](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html#cfn-pseudo-param-partition)

File: setup.py [hunk 7]
```diff
@@ -29,7 +29,7 @@
     # with the LICENSE file added to its root folder
     # and the following patches applied
     # https://github.com/virtuald/pyhcl/pull/57
-    'pyhcl~=0.3',
+    'pyhcl<0.3.14',
```
Comment (troyready) [inline setup.py]: I think a rebase should make dep change go away.

File: src/runway/blueprints/staticsite/staticsite.py [hunk 8]
```diff
@@ -168,16 +360,34 @@ def create_template(self):
...
+        self.template.add_output(Output(
             'BucketName',
             Description='Name of website bucket',
             Value=bucket.ref()
         ))
+        self.template.add_output(Output(
+            'BucketWebsiteURL',
+            Condition="CFDisabled",
+            Description='URL of the bucket website',
+            Value=GetAtt(bucket, 'WebsiteURL')
```
Comment (troyready) [inline src/runway/blueprints/staticsite/staticsite.py]: Please change this to `bucket.get_att('WebsiteURL')`

(doesn't change the generated template at all, really just a style/consistency thing)

File: src/runway/blueprints/staticsite/staticsite.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (voodooGQ) [reply src/runway/blueprints/staticsite/staticsite.py]: There we go that did it. Thank you!

File: src/runway/blueprints/staticsite/staticsite.py (hunk 8 above)
Comment (voodooGQ) [reply src/runway/blueprints/staticsite/staticsite.py]: Updated

File: setup.py (hunk 7 above)
Comment (voodooGQ) [reply setup.py]: Rebased

**Pull request merged**